File: internal/client/conflicts.go
// 冲突副本的发现与处理（M3 管理功能的数据层）。
// 冲突副本是真实文件（FR-S7 命名 `name (conflict from 设备).ext`），
// 因此"处理冲突"就是纯文件操作：保留本地=删副本；采用副本=用副本覆盖原名再删副本。
// 后续同步引擎会自动把结果传播到服务端与其他设备，无需新同步逻辑。
package client

import (
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"strings"
)

const conflictMarker = " (conflict from "

// Conflict 一条待处理的冲突。
type Conflict struct {
	Folder  string `json:"folder"`   // 所属同步文件夹名
	Rel     string `json:"rel"`      // 原始文件的相对路径
	CopyRel string `json:"copy_rel"` // 冲突副本的相对路径
	Size    int64  `json:"size"`
	MTime   int64  `json:"mtime"`
}

// ListConflicts 扫描文件夹根下的冲突副本文件。
// localPath → folderName 的映射由调用方提供（daemon 持有配置）。
func ListConflicts(root, folderName string) ([]Conflict, error) {
	var out []Conflict
	err := filepath.WalkDir(root, func(p string, d fs.DirEntry, err error) error {
		if err != nil {
			return nil
		}
		if d.IsDir() {
			name := d.Name()
			if name == ".y-sync" || name == ".git" || name == ".svn" || name == ".hg" {
				return filepath.SkipDir
			}
			return nil
		}
		base := d.Name()
		i := strings.Index(base, conflictMarker)
		if i < 0 {
			return nil
		}
		// 从文件名还原原始文件名：name (conflict from DEV)[ 2].ext → name.ext
		ext := filepath.Ext(base)
		origName := base[:i] + ext
		rel, err := filepath.Rel(root, p)
		if err != nil {
			return nil
		}
		rel = filepath.ToSlash(rel)
		dir := filepath.ToSlash(filepath.Dir(rel))
		origRel := origName
		if dir != "." {
			origRel = dir + "/" + origName
		}
		info, err := d.Info()
		if err != nil {
			return nil
		}
		out = append(out, Conflict{
			Folder:  folderName,
			Rel:     origRel,
			CopyRel: rel,
			Size:    info.Size(),
			MTime:   info.ModTime().UnixMilli(),
		})
		return nil
	})
	return out, err
}

// ResolveKeepLocal 保留原名文件：删除冲突副本（删除将同步传播）。
func ResolveKeepLocal(root string, c Conflict) error {
	if !strings.Contains(filepath.Base(c.CopyRel), conflictMarker) {
		return fmt.Errorf("%q 不是冲突副本", c.CopyRel)
	}
	return os.Remove(absJoin(root, c.CopyRel))
}

// ResolveKeepCopy 采用副本：副本内容覆盖原名文件，然后删除副本。
func ResolveKeepCopy(root string, c Conflict) error {
	if !strings.Contains(filepath.Base(c.CopyRel), conflictMarker) {
		return fmt.Errorf("%q 不是冲突副本", c.CopyRel)
	}
	copyAbs := absJoin(root, c.CopyRel)
	origAbs := absJoin(root, c.Rel)
	if _, err := os.Stat(copyAbs); err != nil {
		return err
	}
	data, err := os.ReadFile(copyAbs)
	if err != nil {
		return err
	}
	// 写临时文件再改名，与引擎的写入安全约定一致
	tmp, err := os.CreateTemp(filepath.Dir(origAbs), ".ysync-resolve-*")
	if err != nil {
		return err
	}
	tmpName := tmp.Name()
	if _, err := tmp.Write(data); err != nil {
		tmp.Close()
		os.Remove(tmpName)
		return err
	}
	if err := tmp.Close(); err != nil {
		os.Remove(tmpName)
		return err
	}
	if err := os.Rename(tmpName, origAbs); err != nil {
		os.Remove(tmpName)
		return err
	}
	return os.Remove(copyAbs)
}

File: rust/ysyncd/src/conflicts.rs
//! 冲突副本的发现与处理（FR-S7）：与 Go internal/client/conflicts.go 行为一致。
//! 处理是纯文件操作，后续同步自动传播。
use std::path::Path;

use serde::{Deserialize, Serialize};

use crate::engine::abs_join;

pub const CONFLICT_MARKER: &str = " (conflict from ";

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct Conflict {
    #[serde(rename = "folder", default)]
    pub folder: String,
    #[serde(rename = "rel", default)]
    pub rel: String,
    #[serde(rename = "copy_rel", default)]
    pub copy_rel: String,
    #[serde(rename = "size", default)]
    pub size: i64,
    #[serde(rename = "mtime", default)]
    pub mtime: i64,
}

/// 扫描文件夹根下的冲突副本文件。
pub fn list_conflicts(root: &Path, folder_name: &str) -> Vec<Conflict> {
    let mut out = Vec::new();
    walk(root, root, folder_name, &mut out);
    out
}

fn walk(base: &Path, dir: &Path, folder: &str, out: &mut Vec<Conflict>) {
    let Ok(entries) = std::fs::read_dir(dir) else { return };
    let mut entries: Vec<_> = entries.filter_map(|e| e.ok()).collect();
    entries.sort_by_key(|e| e.file_name());
    for e in entries {
        let Ok(ft) = e.file_type() else { continue };
        let p = e.path();
        if ft.is_dir() {
            match e.file_name().to_str() {
                Some(".y-sync") | Some(".git") | Some(".svn") | Some(".hg") => continue,
                _ => {}
            }
            walk(base, &p, folder, out);
            continue;
        }
        let Some(base_name) = e.file_name().to_str().map(|s| s.to_string()) else { continue };
        let Some(i) = base_name.find(CONFLICT_MARKER) else { continue };
        let ext = match base_name.rfind('.') {
            Some(j) if j > i => &base_name[j..],
            _ => "",
        };
        let orig_name = format!("{}{ext}", &base_name[..i]);
        let Ok(rel) = p.strip_prefix(base) else { continue };
        let rel = rel.to_string_lossy().replace('\\', "/");
        let copy_rel = rel.clone();
        let dir_part = match rel.rfind('/') {
            Some(j) => &rel[..j],
            None => "",
        };
        let orig_rel = if dir_part.is_empty() {
            orig_name.clone()
        } else {
            format!("{dir_part}/{orig_name}")
        };
        let (size, mtime) = match e.metadata() {
            Ok(m) => (
                m.len() as i64,
                m.modified()
                    .ok()
                    .and_then(|t| t.duration_since(std::time::UNIX_EPOCH).ok())
                    .map(|d| d.as_millis() as i64)
                    .unwrap_or(0),
            ),
            Err(_) => (0, 0),
        };
        out.push(Conflict {
            folder: folder.to_string(),
            rel: orig_rel,
            copy_rel,
            size,
            mtime,
        });
    }
}

fn is_conflict_copy(name: &str) -> bool {
    name.contains(CONFLICT_MARKER)
}

/// 保留原名文件：删除冲突副本（删除将同步传播）。
pub fn resolve_keep_local(root: &Path, c: &Conflict) -> Result<(), ysync_core::Error> {
    let copy_base = c
        .copy_rel
        .rsplit('/')
        .next()
        .unwrap_or(&c.copy_rel)
        .to_string();
    if !is_conflict_copy(&copy_base) {
        return Err(ysync_core::Error::Msg(format!("{:?} 不是冲突副本", c.copy_rel)));
    }
    std::fs::remove_file(abs_join(root, &c.copy_rel))
        .map_err(|e| ysync_core::Error::Msg(format!("remove: {e}")))
}

/// 采用副本：副本内容覆盖原名文件，然后删除副本。
pub fn resolve_keep_copy(root: &Path, c: &Conflict) -> Result<(), ysync_core::Error> {
    let copy_base = c
        .copy_rel
        .rsplit('/')
        .next()
        .unwrap_or(&c.copy_rel)
        .to_string();
    if !is_conflict_copy(&copy_base) {
        return Err(ysync_core::Error::Msg(format!("{:?} 不是冲突副本", c.copy_rel)));
    }
    let copy_abs = abs_join(root, &c.copy_rel);
    let orig_abs = abs_join(root, &c.rel);
    let data = std::fs::read(&copy_abs).map_err(|e| ysync_core::Error::Msg(format!("read: {e}")))?;
    let tmp = orig_abs.with_file_name(format!(
        ".ysync-resolve-{}",
        std::process::id()
    ));
    std::fs::write(&tmp, &data).map_err(|e| ysync_core::Error::Msg(format!("write: {e}")))?;
    std::fs::rename(&tmp, &orig_abs).map_err(|e| ysync_core::Error::Msg(format!("rename: {e}")))?;
    std::fs::remove_file(&copy_abs).map_err(|e| ysync_core::Error::Msg(format!("remove: {e}")))
}
